blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc

```diff
@@ -0,0 +1,647 @@
++
+Any other messages are enriched with the corresponding transaction metadata via a custom `Transformer` (see below) and finally are written to an output topic
+
+.Topic Retention
+[NOTE]
+===============================
+In order to manage the growth of involved topics, the retention policy for each topic should be well-defined.
+For instance for the actual audit log topic with the enriched change events, a time based retention policy might be suitable, keeping each log event for as long as needed as per your requirements.
+The transaction metadata topic on the other hand can be fairly short-lived, as its entries are not needed any longer, once all corresponding data change events have been processed.
+It may be a good idea to set up some monitoring of the end-to-end lag in order to make sure the log enricher stream application keeps up with the incoming messages and doesn't fall behind that far so it is at risk of transaction messages being discarded before processing the corresponding change events.
+===============================
+
+The topic names are injected using the https://microprofile.io/project/eclipse/microprofile-config[MicroProfile Config API], with the values being provided in Quarkus _application.properties_ configuration file.
+Besides the topic names, this file also has the information about the Kafka bootstrap server, default serdes any more:
+
+[source]
+----
+audit.context.data.topic=dbserver1.inventory.transaction_context_data
+audit.vegetables.topic=dbserver1.inventory.vegetable
+audit.vegetables.enriched.topic=dbserver1.inventory.vegetable.enriched
+
+# may be overridden with env vars
+quarkus.kafka-streams.bootstrap-servers=localhost:9092
+quarkus.kafka-streams.application-id=auditlog-enricher
+quarkus.kafka-streams.topics=${audit.context.data.topic},${audit.vegetables.topic}
+
+# pass-through
+kafka-streams.cache.max.bytes.buffering=10240
+kafka-streams.commit.interval.ms=1000
+kafka-streams.metadata.max.age.ms=500
+kafka-streams.auto.offset.reset=earliest
+kafka-streams.metrics.recording.level=DEBUG
+kafka-streams.default.key.serde=io.debezium.demos.auditing.enricher.JsonObjectSerde
+kafka-streams.default.value.serde=io.debezium.demos.auditing.enricher.JsonObjectSerde
+kafka-streams.processing.guarantee=exactly_once
+----
+
+In the next step let's take a look at the `ChangeEventEnricher` class, our custom transformer.
+The implemention is based on the assumption that change events are serialized as JSON,
+but of course it could be done equally well using other formats such as Avro or Protocol Buffers.
+
+This is a bit of code, but hopefully its decomposition into multiple smaller methods makes it comprehensible:
```

Comment:
Since you describe only the most relevant parts below the code block maybe there is not need to provide all code lines here but instead reduce it to the code blocks that are also discussed in the text and link to the source file. WDYT?

Reply:
Yes, I was wondering that, too. But I've added AsciiDoc callouts now, which I think help a lot with keeping this comprehensible (hard to see here on GitHub, but they'll show up in the rendered output on the website). So I think it's ok as is. I've already removed the `BufferOffsets` class and just point to it on GitHub.

Reply:
Here's an image of the rendered callouts; they're referenced in text below the listing. <img width="859" alt="Screen Shot 2019-09-29 at 22 14 43" src="https://user-images.githubusercontent.com/28612/65838747-ae97be80-e306-11e9-9b45-b7670e31a7e3.png">